nilearn/glm/tests/test_bids.py

```diff
@@ -0,0 +1,202 @@
+"""Test the first level model on BIDS datasets."""
+import os
+import shutil
+
+from pathlib import Path
+
+import pytest
+
+from nilearn._utils.data_gen import (create_fake_bids_dataset)
+from nilearn.glm.first_level import (first_level_from_bids)
+from nilearn.interfaces.bids import get_bids_files
+from nibabel.tmpdirs import InTemporaryDirectory
+
+def test_first_level_from_bids_bug_3029():
+    "Test error when events.tsv is missing for a bold file."
+    with InTemporaryDirectory():
+
+        bids_path = create_fake_bids_dataset(n_sub=2,
+                                            n_ses=2,
+                                            tasks=['main'],
+                                            n_runs=[2])
+        files_to_rename = Path(bids_path).joinpath('derivatives').glob(
+            '**/func/*_task-main_*desc-*')
+        for file in files_to_rename:
+            new_file = file.parent / file.name.replace('run-0', 'run-')
+            file.rename(new_file)
+
+        with pytest.raises(ValueError, 
+                           match=".*events.tsv files.*"):
+            first_level_from_bids(
+                bids_path,
+                task_label='main',
+                space_label='MNI',
+                img_filters=[('desc', 'preproc')],
+                verbose=1)
+
+def test_first_level_from_bids():
+    with InTemporaryDirectory():
+
+        bids_path = create_fake_bids_dataset(n_sub=3,
+                                             n_ses=2,
+                                             tasks=['localizer', 'main'],
+                                             n_runs=[1, 3])
+        # test output is as expected
+        models, m_imgs, m_events, m_confounds = first_level_from_bids(
+            bids_path,
+            task_label='main',
+            space_label='MNI',
+            img_filters=[('desc', 'preproc')])
+        assert len(models) == len(m_imgs)
+        assert len(models) == len(m_events)
+        assert len(models) == len(m_confounds)
+
+
+@pytest.mark.parametrize('entity', ['acq',
+                                    'ce',
+                                    'dir',
+                                    'rec',
+                                    'echo',
+                                    'res',
+                                    'den'])
```

Comment:
If you create the `ENTITIES_RAW` etc. as global variables, you can import them and use them here.

Reply:
yeah except that because of how run is handled in the data generating function I cannot really have a single constant that will work everywhere.